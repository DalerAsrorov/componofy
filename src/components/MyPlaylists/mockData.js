import {
  MOCK_EXTERNAL_URL,
  MOCK_ALBUM,
  MOCK_TRACK,
  MOCK_TRACKS,
  MOCK_IMAGE,
  MOCK_ARTIST,
  MOCK_NAVIGATION,
  MOCK_PLAYLISTS,
} from '../../utils/mocks';

export const MOCK_MY_PLAYLISTS = {
  playlists: MOCK_PLAYLISTS,
  isFetching: false,
  numberOfTracks: MOCK_PLAYLISTS.length,
  currentOffset: 0,
  playlistsRemaining: 0,
  areAllOpen: false,
  canLoadMore: false,
  isVisited: false,
};
