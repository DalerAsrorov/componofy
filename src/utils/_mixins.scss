@mixin vertical-align($position: relative) {
  position: $position;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}
