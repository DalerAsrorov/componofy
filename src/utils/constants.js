import PropTypes from 'prop-types';
import { lightBlue, cyan, green } from '@material-ui/core/colors';

export const EXTERNAL_URLS_PROPTYPE = PropTypes.shape({
  spotify: PropTypes.string.isRequired,
});

export const IMAGE_PROPTYPE = PropTypes.shape({
  url: PropTypes.string.isRequired,
  height: PropTypes.number,
  width: PropTypes.number,
});

export const ARTIST_PROPTYPE = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  external_urls: EXTERNAL_URLS_PROPTYPE.isRequired,
});

export const ALBUM_PROPTYPE = PropTypes.shape({
  artists: PropTypes.arrayOf(ARTIST_PROPTYPE),
  name: PropTypes.string.isRequired,
  external_urls: EXTERNAL_URLS_PROPTYPE.isRequired,
  album_type: PropTypes.string,
  images: PropTypes.arrayOf(IMAGE_PROPTYPE),
});

export const TRACK_PROPTYPE = PropTypes.shape({
  artists: PropTypes.arrayOf(ARTIST_PROPTYPE).isRequired,
  external_urls: EXTERNAL_URLS_PROPTYPE.isRequired,
  track_number: PropTypes.number.isRequired,
  duration_ms: PropTypes.number.isRequired,
  popularity: PropTypes.number.isRequired,
  album: ALBUM_PROPTYPE.isRequired,
  id: PropTypes.string.isRequired,
  preview_url: PropTypes.string,
});

export const PLAYLIST_PROPTYPE = PropTypes.shape({
  id: PropTypes.string.isRequired,
  href: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  external_urls: EXTERNAL_URLS_PROPTYPE.isRequired,
  tracks: PropTypes.shape({ list: PropTypes.array }).isRequired,
  owner: PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
  }).isRequired,
  images: PropTypes.arrayOf(IMAGE_PROPTYPE),
  type: PropTypes.string,
  public: PropTypes.bool,
  // If true, the list of playlist tracks will be expanded
  isOpen: PropTypes.bool,
});

export const USER_PROPTYPE = PropTypes.shape({
  isAuthenticated: PropTypes.bool.isRequired,
  refreshToken: PropTypes.string.isRequired,
  accessToken: PropTypes.string.isRequired,
  sessionID: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
});

export const PLAYLISTS_PROPTYPE = PropTypes.shape({
  playlists: PropTypes.arrayOf(PLAYLIST_PROPTYPE).isRequired,
  isFetching: PropTypes.bool.isRequired,
  numberOfTracks: PropTypes.number.isRequired,
  currentOffset: PropTypes.number.isRequired,
  playlistsRemaining: PropTypes.number.isRequired,
  areAllOpen: PropTypes.bool.isRequired,
  canLoadMore: PropTypes.bool.isRequired,
  isVisited: PropTypes.bool.isRequired,
  suggestedPlaylist: PropTypes.array,
  lastUpdated: PropTypes.number,
});

export const ROUTER_PROPTYPE = PropTypes.shape({
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string,
    hash: PropTypes.string,
  }),
});

export const searchKeyMap = {
  focusSearch: 'ctrl+f',
};

export const MOST_LIGHT_BLUE_COLOR = lightBlue[50];
export const LIGHT_BLUE_COLOR = lightBlue[600];
export const LIGHT_CYAN_COLOR = cyan[500];
export const SUCCESS_COLOR = green[600];

export const menuButtonStyle = {};

export const SCROLL_DURATION = 500;
export const OFFSET_LIMIT = 10;
export const PLAYLIST_OFFSET_LIMIT = 20;

// Max image size is 256KB
// https://developer.spotify.com/web-api/upload-a-custom-playlist-cover-image/
export const MAX_IMAGE_SIZE_LIMIT = 256000;
// make request for refresh token every minute
export const REFRESH_TOKEN_UPDATE_TIME = 60 * 1000;

// YouTube link to Demo
export const DEMO_YOUTUBE_LINK = 'https://www.youtube.com/watch?v=lQnvfRADJMQ';

export const LOAD_MORE_STATUS = {
  // There is no tracks to load
  0: 'All playlists loaded',
  // There is more tracks to load
  1: 'Load more',
  // Is true during data request
  2: 'Loading...',
};

export const SUGGESTED_PLAYLIST_PLACEHOLDER = (tracks = []) => ({
  id: 'suggestedPlaylist',
  href: '#suggestedPlaylist',
  isCustom: true,
  name: 'Componofy Suggested Playlist <3',
  owner: {
    id: '',
    type: '',
  },
  tracks: {
    list: [...tracks],
  },
  external_urls: {
    spotify: '',
  },
});

export const MIN_POPULAR_SCORE = 70;
