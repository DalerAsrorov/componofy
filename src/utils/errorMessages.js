export const NO_TRACK_ID_ERROR =
    'Track with no ID will not be shown and included in final playlist';
