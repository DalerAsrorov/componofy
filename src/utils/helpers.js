import * as R from 'ramda';

const SERVER_URL = 'http://localhost:3001';

export const replaceTo = path => {
    window.location.replace(`${SERVER_URL}${path}`);
};

export const formatPlaylist = unformattedPlaylist => {
    let formatted = unformattedPlaylist;

    return formatted;
};

export const formatTracks = tracks => {
    return tracks.map(trackObject => {
        const { track, ...rest } = trackObject;

        return {
            ...track,
            ...rest
        };
    });
};

export const removeDuplicates = R.curry((data, prop) => {
    return R.pipe(
        R.groupBy(R.prop(prop)),
        R.values,
        R.pipe(R.map(R.take(1)), R.flatten)
    )(data);
});

export const swapKeysAndValues = (map, fn = toString) => {
    return R.zipObj(R.values(map), R.map(fn, R.keys(map)));
};
