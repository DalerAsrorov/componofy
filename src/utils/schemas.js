import { schema } from 'normalizr';

export const track = new schema.Entity('tracks');

export const playlist = new schema.Entity('playlists', {
  tracks: {
    list: [track]
  }
});
