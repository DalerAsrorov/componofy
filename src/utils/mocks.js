import { swapKeysAndValues } from './helpers';

export const MOCK_EXTERNAL_URL = {
  spotify: 'test_external_url'
};

export const ROUTE_INDEX_MAP = {
  0: '/app',
  1: '/app/public',
  2: '/app/componofy'
};

export const MOCK_IMAGE = {
  url: 'test image url',
  height: 500,
  width: 100
};

export const DEFAULT_INDEX = 0;

export const MOCK_NAVIGATION = {
  index: 0,
  currentPage: ROUTE_INDEX_MAP[DEFAULT_INDEX],
  nextIndex: 1,
  nextPage: ROUTE_INDEX_MAP[DEFAULT_INDEX + 1],
  indexToRouteMap: ROUTE_INDEX_MAP,
  routeToIndexMap: swapKeysAndValues(ROUTE_INDEX_MAP, parseInt),
  numberOfPages: 3
};

export const MOCK_ARTIST = {
  id: '324',
  name: 'Test Artist',
  external_urls: MOCK_EXTERNAL_URL
};

export const MOCK_ALBUM = {
  artists: [MOCK_ARTIST],
  name: 'Test album',
  external_urls: MOCK_EXTERNAL_URL,
  album_type: 'test type',
  images: [MOCK_IMAGE, MOCK_IMAGE]
};

export const MOCK_TRACK = {
  id: '1234',
  artists: [MOCK_ARTIST],
  external_urls: MOCK_EXTERNAL_URL,
  track_number: 1,
  duration_ms: 12323,
  popularity: 78,
  album: MOCK_ALBUM
};

export const MOCK_TRACKS = [MOCK_TRACK, MOCK_TRACK, MOCK_TRACK];

export const MOCK_PLAYLIST = {
  id: '8752',
  href: 'testLink',
  name: 'Artist 1',
  external_urls: MOCK_EXTERNAL_URL,
  tracks: {
    list: MOCK_TRACKS
  },
  owner: {
    id: 'testOwner',
    type: 'testType'
  },
  images: [MOCK_IMAGE],
  type: 'testType',
  public: true,
  isOpen: false
};

export const MOCK_PLAYLISTS = [
  {
    id: '8752',
    href: 'testLink',
    name: 'Artist 1',
    external_urls: MOCK_EXTERNAL_URL,
    tracks: {
      list: []
    },
    owner: {
      id: 'testOwner',
      type: 'testType'
    },
    images: [MOCK_IMAGE],
    type: 'testType',
    public: true,
    isOpen: false
  },
  {
    id: '6547',
    href: 'testLink',
    name: 'Artist 1',
    external_urls: MOCK_EXTERNAL_URL,
    tracks: {
      list: []
    },
    owner: {
      id: 'testOwner',
      type: 'testType'
    },
    images: [MOCK_IMAGE],
    type: 'testType',
    public: true,
    isOpen: false
  },
  {
    id: '4234',
    href: 'testLink',
    name: 'Artist 1',
    external_urls: MOCK_EXTERNAL_URL,
    tracks: {
      list: []
    },
    owner: {
      id: 'testOwner',
      type: 'testType'
    },
    images: [MOCK_IMAGE],
    type: 'testType',
    public: true,
    isOpen: false
  }
];
