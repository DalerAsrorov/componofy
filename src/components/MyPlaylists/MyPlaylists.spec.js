import React from 'react';
import Tabs, { Tab } from 'material-ui/Tabs';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createShallow, createMount } from 'material-ui/test-utils';
import MyPlaylsits from './';
import { MOCK_MY_PLAYLISTS } from './mockData';
import { MOCK_NAVIGATION } from '../../utils/mocks';

Enzyme.configure({ adapter: new Adapter() });

const requiredProps = {
  startPlaylistTracksReorderProcess: jest.fn(),
  setOpenStatusForAllPlaylists: jest.fn(),
  myPlaylistsHasOpenPlaylist: false,
  generateSuggestedPlaylists: jest.fn(),
  setOpenStatusMyPlaylists: jest.fn(),
  removePlaylistFromFinal: jest.fn(),
  reorderPlaylistTracks: jest.fn(),
  setPlaylistDragStatus: jest.fn(),
  addPlaylistToFinal: jest.fn(),
  fetchMyPlaylists: jest.fn(),
  myPlaylists: MOCK_MY_PLAYLISTS,
  setPlaylistOpen: jest.fn(),
  setMySearchTerm: jest.fn(),
  addErrorToApp: jest.fn(),
  navigation: MOCK_NAVIGATION,
  logOutUser: jest.fn(),
  classes: {}
};

/**
 * NOTE: Testing this causes a lot of errors
 * will need to re-visit on how to test this huge component
 * properly.
 */
describe('My Playlists', () => {
  let mount, shallow;

  beforeEach(() => {
    mount = createMount();
    shallow = createShallow();
  });

  it('renders with required props', () => {
    let wrapper = shallow(<MyPlaylsits {...requiredProps} />);

    expect(wrapper).toMatchSnapshot();
  });
});
