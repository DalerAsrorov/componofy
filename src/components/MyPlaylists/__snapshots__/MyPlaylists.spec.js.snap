// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Playlists renders with required props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(MyPlaylists)
    addErrorToApp={[Function]}
    addPlaylistToFinal={[Function]}
    classes={Object {}}
    fetchMyPlaylists={[Function]}
    generateSuggestedPlaylists={[Function]}
    logOutUser={[Function]}
    myPlaylists={
      Object {
        "areAllOpen": false,
        "canLoadMore": false,
        "currentOffset": 0,
        "isFetching": false,
        "isVisited": false,
        "numberOfTracks": 3,
        "playlists": Array [
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "8752",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "6547",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "4234",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
        ],
        "playlistsRemaining": 0,
      }
    }
    myPlaylistsHasOpenPlaylist={false}
    navigation={
      Object {
        "currentPage": "/app",
        "index": 0,
        "indexToRouteMap": Object {
          "0": "/app",
          "1": "/app/public",
          "2": "/app/componofy",
        },
        "nextIndex": 1,
        "nextPage": "/app/public",
        "numberOfPages": 3,
        "routeToIndexMap": Object {
          "/app": 0,
          "/app/componofy": 2,
          "/app/public": 1,
        },
      }
    }
    removePlaylistFromFinal={[Function]}
    reorderPlaylistTracks={[Function]}
    setMySearchTerm={[Function]}
    setOpenStatusForAllPlaylists={[Function]}
    setOpenStatusMyPlaylists={[Function]}
    setPlaylistDragStatus={[Function]}
    setPlaylistOpen={[Function]}
    startPlaylistTracksReorderProcess={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "addErrorToApp": [Function],
      "addPlaylistToFinal": [Function],
      "classes": Object {
        "footerPanel": "MyPlaylists-footerPanel-5",
        "hotKeys": "MyPlaylists-hotKeys-3",
        "loadmore": "MyPlaylists-loadmore-1",
        "playlistRemaining": "MyPlaylists-playlistRemaining-4",
        "searchAdortment": "MyPlaylists-searchAdortment-2",
      },
      "fetchMyPlaylists": [Function],
      "generateSuggestedPlaylists": [Function],
      "logOutUser": [Function],
      "myPlaylists": Object {
        "areAllOpen": false,
        "canLoadMore": false,
        "currentOffset": 0,
        "isFetching": false,
        "isVisited": false,
        "numberOfTracks": 3,
        "playlists": Array [
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "8752",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "6547",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
          Object {
            "external_urls": Object {
              "spotify": "test_external_url",
            },
            "href": "testLink",
            "id": "4234",
            "images": Array [
              Object {
                "height": 500,
                "url": "test image url",
                "width": 100,
              },
            ],
            "isOpen": false,
            "name": "Artist 1",
            "owner": Object {
              "id": "testOwner",
              "type": "testType",
            },
            "public": true,
            "tracks": Object {
              "list": Array [],
            },
            "type": "testType",
          },
        ],
        "playlistsRemaining": 0,
      },
      "myPlaylistsHasOpenPlaylist": false,
      "navigation": Object {
        "currentPage": "/app",
        "index": 0,
        "indexToRouteMap": Object {
          "0": "/app",
          "1": "/app/public",
          "2": "/app/componofy",
        },
        "nextIndex": 1,
        "nextPage": "/app/public",
        "numberOfPages": 3,
        "routeToIndexMap": Object {
          "/app": 0,
          "/app/componofy": 2,
          "/app/public": 1,
        },
      },
      "removePlaylistFromFinal": [Function],
      "reorderPlaylistTracks": [Function],
      "setMySearchTerm": [Function],
      "setOpenStatusForAllPlaylists": [Function],
      "setOpenStatusMyPlaylists": [Function],
      "setPlaylistDragStatus": [Function],
      "setPlaylistOpen": [Function],
      "startPlaylistTracksReorderProcess": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "addErrorToApp": [Function],
        "addPlaylistToFinal": [Function],
        "classes": Object {
          "footerPanel": "MyPlaylists-footerPanel-5",
          "hotKeys": "MyPlaylists-hotKeys-3",
          "loadmore": "MyPlaylists-loadmore-1",
          "playlistRemaining": "MyPlaylists-playlistRemaining-4",
          "searchAdortment": "MyPlaylists-searchAdortment-2",
        },
        "fetchMyPlaylists": [Function],
        "generateSuggestedPlaylists": [Function],
        "logOutUser": [Function],
        "myPlaylists": Object {
          "areAllOpen": false,
          "canLoadMore": false,
          "currentOffset": 0,
          "isFetching": false,
          "isVisited": false,
          "numberOfTracks": 3,
          "playlists": Array [
            Object {
              "external_urls": Object {
                "spotify": "test_external_url",
              },
              "href": "testLink",
              "id": "8752",
              "images": Array [
                Object {
                  "height": 500,
                  "url": "test image url",
                  "width": 100,
                },
              ],
              "isOpen": false,
              "name": "Artist 1",
              "owner": Object {
                "id": "testOwner",
                "type": "testType",
              },
              "public": true,
              "tracks": Object {
                "list": Array [],
              },
              "type": "testType",
            },
            Object {
              "external_urls": Object {
                "spotify": "test_external_url",
              },
              "href": "testLink",
              "id": "6547",
              "images": Array [
                Object {
                  "height": 500,
                  "url": "test image url",
                  "width": 100,
                },
              ],
              "isOpen": false,
              "name": "Artist 1",
              "owner": Object {
                "id": "testOwner",
                "type": "testType",
              },
              "public": true,
              "tracks": Object {
                "list": Array [],
              },
              "type": "testType",
            },
            Object {
              "external_urls": Object {
                "spotify": "test_external_url",
              },
              "href": "testLink",
              "id": "4234",
              "images": Array [
                Object {
                  "height": 500,
                  "url": "test image url",
                  "width": 100,
                },
              ],
              "isOpen": false,
              "name": "Artist 1",
              "owner": Object {
                "id": "testOwner",
                "type": "testType",
              },
              "public": true,
              "tracks": Object {
                "list": Array [],
              },
              "type": "testType",
            },
          ],
          "playlistsRemaining": 0,
        },
        "myPlaylistsHasOpenPlaylist": false,
        "navigation": Object {
          "currentPage": "/app",
          "index": 0,
          "indexToRouteMap": Object {
            "0": "/app",
            "1": "/app/public",
            "2": "/app/componofy",
          },
          "nextIndex": 1,
          "nextPage": "/app/public",
          "numberOfPages": 3,
          "routeToIndexMap": Object {
            "/app": 0,
            "/app/componofy": 2,
            "/app/public": 1,
          },
        },
        "removePlaylistFromFinal": [Function],
        "reorderPlaylistTracks": [Function],
        "setMySearchTerm": [Function],
        "setOpenStatusForAllPlaylists": [Function],
        "setOpenStatusMyPlaylists": [Function],
        "setPlaylistDragStatus": [Function],
        "setPlaylistOpen": [Function],
        "startPlaylistTracksReorderProcess": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
